docs/src/tutorials/write-my-software-management-plugin.md

```diff
@@ -0,0 +1,374 @@
+
+Here is the table of the commands that you can use in a plugin.
+
+|Command|Input arguments|Expected output|Description|
+|---|---|---|---|
+|list| - | JSON |Returns the list of software modules that have been installed with this plugin.|
+|prepare| - | - |Executes the provided actions before a sequence of install and remove commands.|
+|finalize| - | - |Executes the provided actions after a sequence of install and remove commands.|
+|install| NAME [--version VERSION] [--file FILE] | - |Executes the action of installation.|
+|remove| NAME [--version VERSION] | - |Executes the action of uninstallation.|
+|update-list| COMMAND NAME [--version VERSION] [--file FILE] | - |Executes the list of `install` and `remove` commands.|
+
+The order of the commands invoked by the Software Management Agent is:
+`prepare` -> `update-list` or [`install`, `remove`] ->`finalize`
+
+> Note: There is no guarantee of the order between `install` and `remove`.
+> If you need a specific order, use `update-list` command instead.
+
+In the following sections, we will dive into each command and other rules deeply. 
+
+## Input, Output, and Errors
+
+Before we dive into each command, we should clarify the basic rules of plugins.
+
+### Input
+
+The command arguments themselves and further arguments must be given from **stdin**.
+
+### Output
+
+The **stdout** and **stderr** of the process running a plugin command are captured by the Software Management Agent.
+
+### Exit status
+
+The exit status of plugins are interpreted by sm-agent as follows:
+- **0**: success.
+- **1**: usage. The command arguments cannot be interpreted, and the command has not been launched.
+- **2**: failure. The command failed and there is no point to retry.
+- **3**: retry. The command failed but might be successful later (for instance, when the network will be back).
+
+## List
+
+The `list` command is responsible to return the list of the installed software modules.
+
+Rules:
+
+- This command takes no arguments.
+- The output must be in JSON format including:
+  - **name**: the name of the software module, e.g. `mosquitto`.
+  This name is the name that has been used to install it and that needs to be used to remove it.
+  - **version**: the version currently installed.
+  This is a string that can only be interpreted in the context of the plugin.
+  
+  
+Given that your plugin is named `myplugin`, the image name is `mosquitto`, and the version is `1.5.7-1+deb10u1`,
+then the Software Management plugin calls
+
+```shell
+myplugin list
+```
+
+to report the list of software modules installed. `myplugin` should output
+
+```json
+{
+   "name":"mosquitto",
+   "version":"1.5.7-1+deb10u1"
+}
+```
+
+with exit code `0` (successful).
+
+A plugin must return this JSON structure per software module.
+In the _docker_ file example, the following command outputs such JSON structures.
+
+```shell
+sudo docker image list --format '{"name":"{{.Repository}}","version":"{{.Tag}}"}'
+```
+
```

Comment:
Here is a good place to add the output of `$ docker list` as an example of jsonlines expected output